python/cudf/cudf/core/scalar.py

```diff
@@ -1,26 +1,208 @@
+                raise TypeError(
+                    "dtype required when constructing a null scalar"
+                )
+        else:
+            dtype = value.dtype
+
+    if not isinstance(dtype, cudf.core.dtypes.DecimalDtype):
+        dtype = cudf.dtype(dtype)
+
+    if not valid:
+        value = NaT if dtype.kind in "mM" else NA
+
+    return value, dtype
+
+
+def _replace_nested(obj, check, replacement):
+    if isinstance(obj, list):
+        for i, item in enumerate(obj):
+            if check(item):
+                obj[i] = replacement
+            elif isinstance(item, (dict, list)):
+                _replace_nested(item, check, replacement)
+    elif isinstance(obj, dict):
+        for k, v in obj.items():
+            if check(v):
+                obj[k] = replacement
+            elif isinstance(v, (dict, list)):
+                _replace_nested(v, check, replacement)
+
+
+def _maybe_nested_pa_scalar_to_py(pa_scalar: pa.Scalar) -> Any:
+    """
+    Convert a "nested" pyarrow scalar to a Python object.
+
+    These scalars come from pylibcudf.Scalar where field names can be
+    duplicate empty strings.
+
+    Parameters
+    ----------
+    pa_scalar: pa.Scalar
+
+    Returns
+    -------
+    Any
+        Python scalar
+    """
+    if not pa_scalar.is_valid:
+        return pa_scalar.as_py()
+    elif pa.types.is_struct(pa_scalar.type):
+        return {
+            str(i): _maybe_nested_pa_scalar_to_py(val)
+            for i, (_, val) in enumerate(pa_scalar.items())
+        }
+    elif pa.types.is_list(pa_scalar.type):
+        return [_maybe_nested_pa_scalar_to_py(val) for val in pa_scalar]
+    else:
+        return pa_scalar.as_py()
+
+
+def _to_plc_scalar(value: ScalarLike, dtype: Dtype) -> plc.Scalar:
+    """
+    Convert a value and dtype to a pylibcudf Scalar for device-side cudf.Scalar
+
+    Parameters
+    ----------
+    value: Scalarlike
+    dtype: dtypelike
+
+    Returns
+    -------
+    plc.Scalar
+    """
+    if cudf.utils.utils.is_na_like(value):
+        value = None
+    else:
+        # TODO: For now we deepcopy the input value for nested values to avoid
+        # overwriting the input values when replacing nulls. Since it's
+        # just host values it's not that expensive, but we could consider
+        # alternatives.
+        if isinstance(value, (list, dict)):
+            value = copy.deepcopy(value)
+        _replace_nested(value, cudf.utils.utils.is_na_like, None)
+
+    if isinstance(dtype, cudf.core.dtypes._BaseDtype):
+        pa_type = dtype.to_arrow()
+    elif pd.api.types.is_string_dtype(dtype):
+        # Have to manually convert object types, which we use internally
+        # for strings but pyarrow only supports as unicode 'U'
+        pa_type = pa.string()
+    else:
+        pa_type = pa.from_numpy_dtype(dtype)
+
+    pa_scalar = pa.scalar(value, type=pa_type)
+    plc_scalar = plc.interop.from_arrow(pa_scalar)
+    if isinstance(dtype, (Decimal32Dtype, Decimal64Dtype)):
```

Comment:
Keep an eye #17422 since that will necessitate corresponding changes on the Python side. I don't think anything will break if we don't change anything we'll just be able to avoid these types of conversions afterwards.

Reply:
Nice! Yes it would be great to avoid this